File: Data Structures & Algorithms/climbing-stairs/submission-3.java
class Solution {

    public static int helper(int[] dp,int curr){
        if(curr<=1){
            return 1;
        }
        if(dp[curr]!=0){
            return dp[curr];
        }
        int result=helper(dp,curr-1)+helper(dp,curr-2);
        dp[curr]=result;
        return result;

    }

    public int climbStairs(int n) {
        if(n==1){
            return 1;
        }
        int[] dp=new int[n+1];
        dp[0]=0;
        dp[1]=1;
        return helper(dp,n);
    }
}

File: Data Structures & Algorithms/permutations/submission-3.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result=[]

        n=len(nums)
        res=[]
        s=set()
        def helper():
            if len(res)==n:
                result.append(res[:])
                return
            for i in nums:
                if i in s:
                    continue
                res.append(i)
                s.add(i)
                helper()
                s.remove(i)
                res.pop()
        helper()
        return result